typescript-loader/loader.js

```diff
@@ -0,0 +1,73 @@
+// 💡 It's probably a good idea to put this loader near the end of the chain.
+
+import path from 'node:path';
+
+import { transform } from 'esbuild';
+
+const jsToTs = new Map([
+  // Sorted by priority (first wins)
+  ['.js',  '.ts' ],
+  ['.cjs', '.cts'],
+  ['.mjs', '.mts'],
+  ['.jsx', '.tsx'],
+]);
+const jsExts = new Set(jsToTs.keys());
+const tsExts = new Set(jsToTs.values());
+
+export function resolve(specifier, context, nextResolve) {
+  const { base, ext, name } = path.parse(specifier);
+
+  if (tsExts.has(ext)) { // No guessing needed
+    const { url } = nextResolve(specifier, context);
+
+    return {
+      format: 'typescript', // Provide a signal to `load`
+      shortCircuit: true,
+      url,
+    };
+  }
+
+  if (!jsExts.has(ext)) { // When file is not ts or js, it's irrelevant
+    return nextResolve(specifier, context);
+  }
+
+  try { // Check whether such a js file does exist
+    return nextResolve(specifier, context);
+  } catch (err) {
+    if (err?.code !== "ERR_MODULE_NOT_FOUND") { throw err; }
+  }
+
+  for (const tsExt of tsExts) { // Finally, check whether any ts file exists
+    try {
+      const maybePath = specifier.replace(base, `${name}${tsExt}`);
```

Comment:
```suggestion
      const maybePath = specifier.replace(ext, tsExt);
```

Reply:
Hum that would replace the first occurrence of `ext`, which is not guaranteed to work (e.g. if the user tries to import `./file.js/index.js`). Maybe this would work?
```suggestion
      const maybePath = specifier.endsWith(ext) ? specifier.slice(-ext.length) + tsExt :
                                                  specifier.replace(`${ext}?`, `${tsExt}?`);
```

Reply:
@GeoffreyBooth I think we should include such a utility in our planned helpers.

Reply:
How robust do you want this to be?  Naive replacement within the specifier will not work in a lot of situations, but I'm not sure whether we should talk about that or if it'll be wasted effort.